sematic/log_reader.py

```diff
@@ -0,0 +1,429 @@
+    more_after: bool
+    lines: List[str]
+    continuation_cursor: Optional[str]
+    log_unavaiable_reason: Optional[str] = None
+
+
+@dataclass
+class Cursor:
+    """A cursor representing a particular place in the process of traversing logs.
+
+    Attributes
+    ----------
+    source_log_key:
+        The storage key for the log that was being used when the search left off. If no
+        logs have been found yet, will be None
+    source_file_line_index:
+        The line number BEFORE filters are applied within the log file being read.
+        It will be the first line that HASN'T yet been read. If no logs have been found,
+        will be -1.
+    filter_strings:
+        The fillter strings that were used for this log traversal.
+    run_id:
+        The run id that was being used for this log traversal.
+    """
+
+    # Why include source log file? Because we will soon likely want to break up
+    # the logs for a single run such that each file contains a *different*
+    # portion of the logs, and we will need to know which file to go to in
+    # order to pick back up. The alternative would be to require
+    # re-traversing already traversed files when continuing.
+    source_log_key: Optional[str]
+    source_file_line_index: int
+    filter_strings: List[str]
+    run_id: str
+
+    def to_token(self) -> str:
+        return str(
+            base64.b64encode(json.dumps(asdict(self)).encode("utf8")), encoding="utf8"
+        )
+
+    @classmethod
+    def from_token(cls, token: str):
+        kwargs = json.loads(
+            base64.b64decode(bytes(token, encoding="utf8")).decode("utf8")
+        )
+        return Cursor(**kwargs)
+
+    @classmethod
+    def nothing_found(cls, filter_strings: List[str], run_id: str):
+        return Cursor(
+            source_log_key=None,
+            source_file_line_index=-1,
+            filter_strings=filter_strings,
+            run_id=run_id,
+        )
+
+
+@dataclass
+class LogLine:
+    source_file: str
+    source_file_index: int
+    line: str
+
+
+def load_log_lines(
+    run_id: str,
+    continuation_cursor: Optional[str],
+    max_lines: int,
+    filter_strings: Optional[List[str]] = None,
+) -> LogLineResult:
+    """Load a portion of the logs for a particular run
+
+    Parameters
+    ----------
+    run_id:
+        The id of the run to get logs for
+    continuation_cursor:
+        A cursor indicating where to continue reading logs from. Should be
+        None if the logs are being read from the beginning.
+    max_lines:
+        The highest number of log lines that should be returned at once
+    filter_strings:
+        Only log lines that contain ALL of the strings in this list will
+        be included in the result
+
+    Returns
+    -------
+    A subset of the logs for the given run
+    """
+    run = get_run(run_id)
+    run_state = FutureState[run.future_state]  # type: ignore
+    still_running = run_state.is_terminal() or run_state == FutureState.RAN
```

Comment:
🤔 `is_terminal` -> `still_running`?

Reply:
you're right, the wording of this is opposite what it represents. Fixing